Add preliminary documentation for rubrics (#9484)
PrairieLearn/PrairieLearn: docs/manualGrading/index.md

> After students have completed their assessments, open the "Manual Grading" tab. This page will list all questions in this assessment, as well as the number of submissions that require grading. For questions that have at least one submission tagged for grading, the "Grade next submission" button will be shown, which will jump directly to one of the submissions that need to be graded. Alternatively, clicking on the question's title will open a list of all student answers for that question.

Reviewer: Ahh okay - that makes sense!

### Setting a rubric for manual grading

An instructor may set a structured rubric to be used in the manual grading process. The rubric allows graders to select among a set of predetermined items with associated points and feedback, and provides a mechanism to facilitate consistency in grading results.

A rubric is composed of a set of rubric items. Each item is assigned a set of points, either positive (representing a credit), negative (representing a penalty) or zero (representing feedback that does not affect the score). A grader can then select which items of the rubric apply to a particular submission among the list. The submission's score is then based on the sum of all the points for all items selected for the submission.

To set a rubric, open an individual question instance. On the grading panel, click on the "Rubric" button to open rubric settings.

![Rubric button on grading panel](rubric_button.png)

![Rubric settings dialog](rubric_settings.png)

For questions that include an auto grading component (with or without a combination of manual points), the rubric may be applied only to manual points or to total points.

> - If it are set to manual points, the score obtained by the rubric grading is added to the points from the auto grader to result in the final score. This type of grading is often used for questions where the manual grading complements the auto grading, such as in grading rationales for specific calculations, or for code style in programming questions.

Reviewer 2: "If it are" should be fixed.

Author: Fixed in 387e2e6.

Each rubric item is set a number of points and a description. This description is shown to both graders and in feedback to students. It is also possible to add an additional detailed explanation to the rubric item, which is shown to students upon request. Finally, a grader note can be set that is visible to graders, but not to students, and can be used to provide instructions on scenarios when the rubric should be used or not.

Items can also be set to always be listed for students (default), or only if it is is selected. This option may be used in cases such as penalties that apply to a small subset of students, or for credit assigned to unconventional correct answers that are not to be advertised to students that did not use them.

In the description, explanation, and grader note, it is possible to use mustache patterns based on the parameters, correct answers and submitted answers for a submission. This will cause the description of the item provided to graders and students to match the question variant assigned to the student, as shown in the screenshot below.

![Rubric items in settings](rubric_items_setting.png)

> This will result in a grading panel containing a description with the value of the parameter replaced with the appropriate value for the student's variant.

Reviewer: ```suggestion
The teaching team will also see a preview of the rubric item description with the appropriate value for that student's submission.
```

I suggest swapping the order of what staff sees and what the student sees (since this is super cool - and most useful for students)

![Rubric items in grading panel](grading_panel_with_rubrics.png)

> Student's feedback is also replaced in a similar manner.

Reviewer: ```suggestion
Using mustache patterns in a rubric item description as shown above will also result in each student seeing a customized description with the parameter replaced with the appropriate value for that student's variant!
```

Reviewer: Swap order of this and staff view.

![Rubric items in student feedback](student_feedback_with_rubrics.png)

> As grading progresses, new scenarios may be found by graders that require changes in the rubric, such as new types of penalties or credits or specific kinds of feedback. The rubric may be changed as required, even after grading is already underway. These changes may affect existing submissions:

Reviewer: ```suggestion
Rubrics in PrairieLearn are dynamic. As grading progresses, new scenarios may be found by graders that require changes in the rubric, such as new types of penalties or credits or more specialized feedback. The rubric may be changed as needed even after grading is already underway - all graders will see changes to the rubric in real time. 

Below is a list of how changes to the rubric will affect affect submissions that have already been manually graded:
```

Reviewer: > "all graders will see changes to the rubric in real time"

Is this true? I can't remember if they need to refresh, or move to a new submission, or if it will alert them to a change.

Author: > Is this true? I can't remember if they need to refresh, or move to a new submission, or if it will alert them to a change.

Other than the grader that updated the rubric, other graders need to refresh.

```suggestion
### Updating the rubric

As grading progresses, new scenarios may be found by graders that require changes in the rubric, such as new types of penalties or credits or more specialized feedback. The rubric may be changed as required, even after grading is already underway. Changes in the rubric are presented to all graders in any submission that still requires grading.

If the rubric is changed after some submissions have already been manually graded, the changes may affect the previously graded submissions, as follows:
```

- Adding new rubric items does not affect the score of existing submissions;

> - Changing the points assigned to individual rubric items will cause any existing submission gradings that use the rubric item to be recalculated with the new points assigned to the items;

Reviewer: ```suggestion
- Changing the point value assigned to a rubric item will trigger a recalculation of the score for all submissions where that rubric item was applied (with the updated point value).
```

Author: Rephrasing to make punctuation consistent and avoid brackets:

```suggestion
- Changing the point value assigned to a rubric item will trigger a recalculation of the score for all submissions where that rubric item was applied, using the updated point value;
```

> - Changing the description or explanation of individual rubric items will update the feedback provided to students, but not the score;

Reviewer: ```suggestion
- Changing the description or explanation of a rubric item will update the feedback shown to students, but will not trigger a score recalculation;
```

RE: 

> but will not trigger a score recalculation;

(personally, I think this part seems a bit silly to even say, but I'm sure there must be some edge case I'm not thinking of where this is important)

> - Deleting rubric items will cause existing gradings that use the rubric item to be recalculated without the item;

Reviewer: ```suggestion
- Deleting a rubric item will trigger a recalculation of the score without the deleted rubric item for all submissions where that rubric item was applied.
```

> - Changing the minimum score or maximum extra credit will cause all existing rubric gradings to be recalculated.

Reviewer: ```suggestion
- Changing the minimum score or maximum extra credit will trigger a score recalculation for all submissions where any rubric item was applied
```

RE:
> trigger a score recalculation for all submissions where any rubric item was applied

I think this is true right? It'll only recalculate the score for submissions where any rubric item was applied.

Author: No, this is for all submissions.

> When updating rubric settings, a flag at the bottom of the settings dialog box gives graders the option of tagging all submissions for manual grading. This option may be used in cases where the changes are significant enough that they warrant a review of all previously graded questions to ensure they take the new rubric into account. This is encouraged in cases where rubric items applied to multiple rubric gradings are deleted.

Reviewer: ```suggestion
When updating rubric settings, a flag at the bottom of the settings dialog box gives graders the option of tagging all submissions (with that rubric item applied) for manual grading. This option may be used in cases where the changes are significant enough to warrant a review of all previously graded questions to ensure the new rubric is taken into account. Substantial changes to rubric item descriptions would normally require at least a spot-check of previously graded submissions.
```

That last sentence was confusing to me...consider rephrasing or removing?

Author: Your sentence is incorrect. Tagging all submissions for new grading applies to all submissions, not only to those with the items changed.

```suggestion
When updating rubric settings, a flag at the bottom of the settings dialog box gives graders the option of tagging all submissions for manual grading. This option may be used in cases where the changes are significant enough to warrant a review of all previously graded questions to ensure they take the new rubric into account. This is encouraged in cases where substantial changes are made, which may require at least a spot-check of previously graded submissions.
```

> To grade a submission, a grader must select the items that apply to a particular submission and press Submit. Additionally, each rubric item (or the first 10 items, in long rubrics) is assigned a keyboard shortcut corresponding to a digit; pressing that digit on the keyboard causes that item to be selected (or cleared, if it is selected).

Reviewer: ```suggestion

### Applying rubric items to submissions

To manually grade a submission with rubrics, a grader must select the items that apply to a particular submission and press Submit. Additionally, each rubric item (or the first 10 items, in long rubrics) is assigned a keyboard shortcut corresponding to a digit; pressing that digit on the keyboard causes that item to be selected (or cleared, if it was already selected).
```

I feel like this section requires a new sub-heading.

Author: Good point, though I'll change the title.

```suggestion
### Grading a submission using a rubric

To manually grade a submission using a rubric, a grader must select the items that apply to a particular submission and press Submit. Additionally, each rubric item (or the first 10 items, in long rubrics) is assigned a keyboard shortcut corresponding to a digit; pressing that digit on the keyboard causes that item to be selected (or cleared, if it is selected).
```

> Graders have the flexibility to apply a grading adjustment (positive or negative) on top of the rubric item. This may be used where instructors need to make judgement calls for unconventional submissions that don't apply to specific rubric items. This option is discouraged in cases where the issue is found in multiple submissions, as additional rubric items provide more flexibility in providing feedback and adjusting grades.

Reviewer: Perhaps a screenshot of this might be useful ? this option is actually not that discoverable. Two options:

<img width="542" alt="Screenshot 2024-03-02 at 12 36 01 AM" src="https://github.com/PrairieLearn/PrairieLearn/assets/2507459/900170f4-f955-4c58-b2cb-db8571af0deb">

<img width="548" alt="Screenshot 2024-03-02 at 12 36 24 AM" src="https://github.com/PrairieLearn/PrairieLearn/assets/2507459/c431ad48-478b-459f-97c5-12fd7cee0789">

Author: I think adding a screenshot here may encourage this feature more than it's supposed to. I want to document that it exists, but not encourage its use extensively.

![Rubric button on grading panel](rubric_button.png)

![Rubric settings dialog](rubric_settings.png)

In questions that only have manual grading, the rubric assigns points to the question as a whole. In autograded questions and questions with a combination of manual and automated grading, the rubric may be set to be applied only to manual points or to total points. This setting causes the grading to be computed as follows:

- If "Apply rubric to manual points" is selected, the score obtained by the rubric grading is added to the points from the auto grader to result in the final score. This type of grading is often used for questions where the manual grading complements the auto grading, such as code style in programming questions, or explanations and rationales for specific calculations.
- If "Apply rubric to total points" is selected, the auto grader component is ignored, and the rubric grading overrides the final score of the question as a whole. This setting is used when the auto grading method is unable to capture all possible correct answers, and manual grading is used to review cases that were not properly graded.

> Instructors may specify the rubric to use positive or negative grading. In positive grading, submissions start at zero, and individual rubric items correspond to students "earning" points that add up to the final score. In negative grading, submissions start at the maximum score, and individual rubric items correspond to penalties or deductions that are subtracted from the score. Both options allow the creation of items with a `0` point value, which provides feedback that does not affect their score. Even within a positive or negative grading rubric, individual rubric items may have positive or negative point values to provide graders with additional flexibility.

Author: ```suggestion
Instructors may set the rubric to use positive or negative grading. In positive grading, submissions start at zero, and individual rubric items correspond to students "earning" points that add up to the final score. In negative grading, submissions start at the maximum score, and individual rubric items correspond to penalties or deductions that are subtracted from the score. Both options allow the creation of items with a `0` point value, which provides feedback that does not affect their score. Even within a positive or negative grading rubric, individual rubric items may have a combination of positive or negative point values to provide graders with additional flexibility.
```

By default, the score produced by the rubric is capped at the maximum points associated to the question, and penalties do not cause the points to get below zero. Optionally, this limit can be extended by assigning a minimum rubric score (typically to allow negative scores) and maximum extra credit to the rubric (to allow points above the maximum).

At minimum, each rubric item consists of a point-value and a description. This description is shown to both graders and in feedback to students. It is also possible to add a more detailed explanation of the rubric item for students to get even more feedback. A grader note can be set that is visible to graders, but not to students, and can be used to provide instructions on scenarios when/how the rubric item should be used (or not).

> By default, all rubric items are set to be shown to students, with visual indicators showing which of the rubric items are applied. Graders also have the option to only show applied rubric items. This option may be useful in specific cases where specific penalties or credit/bonus points apply only to a small subset of students, so they are not shown to everyone to avoid confusion.

Author: ```suggestion
By default, all rubric items are set to be shown to students in all cases, with visual indicators showing which of the rubric items are applied to an individual student. The settings also allow graders to only show applied rubric items. This option may be useful in cases where specific penalties or credit/bonus points apply only to a small subset of students, which may be confusing for students that do not have the rubric item selected.
```

> The description, explanation, and grader note of a rubric item may include mustache patterns that will be replaced in the grading interface and student feedback with appropriate values. These patterns are based on the parameters and correct answers set for the student's variant, or the student's submitted answer. A list of patterns that are available for a particular question are presented at the bottom of the rubric settings dialog for reference. For example, if a variant sets a value to `data["params"]["starting_index"]` in `generate`, a rubric item may include the pattern `{{params.starting_index}}` in the description. This pattern will be replaced with the value of this parameter for the student's specific variant, as seen in the screenshots below.

Author: ```suggestion
The description, explanation, and grader note of a rubric item may include mustache patterns that will be replaced in the grading interface and student feedback with appropriate values. These patterns are based on the parameters and correct answers set for the student's variant, or the student's submitted answer. A list of patterns that are available for a particular question is presented at the bottom of the rubric settings dialog for reference. For example, if a variant sets a value to `data["params"]["starting_index"]` in `generate`, a rubric item may include the pattern `{{params.starting_index}}` in the description, explanation or grader note. This pattern will be replaced with the value of the corresponding parameter for the student's specific variant, as seen in the screenshots below.
```

![Rubric items in settings](rubric_items_setting.png)

> Graders will then see, in the grading panel, the rubric item description updated to reflect the appropriate value of the parameter for the student's variant.

Author: ```suggestion
Graders will then see, in the grading panel, the rubric item description or explanation updated to reflect the appropriate value of the parameter for the student's variant.
```

As grading progresses, new scenarios may be found by graders that require changes in the rubric, such as new types of penalties or credits or more specialized feedback. The rubric may be changed as required, even after grading is already underway. Changes in the rubric are presented to all graders in any submission that still requires grading.

If the rubric is changed after some submissions have already been manually graded, the changes may affect the previously graded submissions, as follows:

- Adding new rubric items does not affect the score of existing submissions;
- Changing the point value assigned to a rubric item will trigger a recalculation of the score for all submissions where that rubric item was applied, using the updated point value;
- Changing the description or explanation of a rubric item will update the feedback shown to students, but will not trigger a score recalculation;
- Deleting a rubric item will trigger a recalculation of the score without the deleted rubric item for all submissions where that rubric item was applied;
- Changing the minimum score or maximum extra credit will cause all existing rubric gradings to be recalculated.

> When updating rubric settings, a flag at the bottom of the settings dialog box gives graders the option of tagging all submissions for manual grading. This option may be used in cases where the changes are significant enough to warrant a review of all previously graded questions to ensure they take the new rubric into account. This is encouraged in cases where substantial changes are made, which may require at least a spot-check of previously graded submissions.

Author: ```suggestion
When updating rubric settings, a flag at the bottom of the settings dialog box gives graders the option of tagging all submissions for manual grading. This option may be used in cases where the changes are significant enough to warrant a review or "spot-check" of all previously graded questions, ensuring they take the new rubric into account.
```

### Grading a submission using a rubric

> To manually grade a submission using a rubric, a grader must select the items that apply to a particular submission and press Submit. Additionally, each rubric item (or the first 10 items, in long rubrics) is assigned a keyboard shortcut corresponding to a digit; pressing that digit on the keyboard causes that item to be selected (or cleared, if it is selected).

Author: ```suggestion
To manually grade a submission using a rubric, a grader must select the items that apply to a particular submission and press Submit. Additionally, each rubric item (or each of the first 10 items, in long rubrics) is assigned a keyboard shortcut corresponding to a digit; pressing that digit on the keyboard causes that item to be selected (or cleared, if it is selected).
```
